agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -2927,40 +2927,64 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
         }
     }
 
-    // The tlv1905NeighborDevice we construct here is not entirely correct, but close enough.
-    // We should create a tlv1905NeighborDevice for every local interface on which a 1905 neighbor
-    // is visible, and then add a list of neighbors visible on that interface. However, since we
-    // don't keep track of the incoming interface of the received topology discovery messages, we
-    // can't do that. Instead, we model it as a single interface, the bridge interface, behind which
-    // all neighbors are visible, with a bridge between them. This is accurate enough to allow us to
-    // work in practice.
-    auto tlv1905NeighborDevice = cmdu_tx.addClass<ieee1905_1::tlv1905NeighborDevice>();
-    if (!tlv1905NeighborDevice) {
-        LOG(ERROR) << "addClass ieee1905_1::tlv1905NeighborDevice failed, mid=" << std::hex
-                   << (int)mid;
-        return false;
-    }
-    tlv1905NeighborDevice->mac_local_iface() = tlvf::mac_from_string(bridge_info.mac);
+    /**
+     * Add a 1905.1 neighbor device TLV for each local interface for which this management entity
+     * has inferred the presence of a 1905.1 neighbor device. Include each discovered neighbor
+     * device in its corresponding 1905.1 neighbor device TLV.
+     *
+     * First, group known 1905 neighbor devices by the local interface that links to them. Create
+     * a map which key is the name of the local interface and the value is the list of neighbor
+     * devices inferred from that interface.
+     */
+    std::unordered_map<std::string, std::vector<sNeighborDevice>> neighbor_devices_by_local_iface;
+    for (const auto &entry : m_1905_neighbor_devices) {
+        const auto &neighbor_device = entry.second;
+        auto &neighbor_devices      = neighbor_devices_by_local_iface[neighbor_device.if_name];
 
-    if (!tlv1905NeighborDevice->alloc_mac_al_1905_device(m_1905_neighbor_devices.size())) {
-        LOG(ERROR) << "alloc_mac_al_1905_device() has failed";
-        return true;
+        neighbor_devices.emplace_back(neighbor_device);
```

Comment:
Well... I admit... I tried and failed.
I feel uncomfortable with building "local structures" and then `transforming` between them whenever there is a need. This is opening a can of possible bugs and creates confusion, uncertainty and repeats.
In this case both `m_1905_neighbor_devices` and `neighbor_devices_by_local_iface` are local structures in the sense that they are pure, native, std structures.
A more robust approach, IMO, would be to have:
```cpp
class NeighborDevices { 
/* construct, fill, query */
// and also -
temaplte<class outIt> get_all_nieghbor_for_ifname(const std::string &if_name, OutIt out);
};
```
Or anything alike.
The idea is this -
Instead of letting the developer of the algorithm mass around with pure structures **when implementing the logical flow** - hide these searches and transformation in a separate place, not in the main flow, not where the logic is. Transformation, searches, validity checks, they all must be moved elsewhere, leaving the flow clean, expressive and accurate.